tool

```diff
@@ -0,0 +1,275 @@
+			-version  show version and exit
+
+			-d        display diffs instead of rewriting files
+			-e        report all errors (not just the first 10 on different lines)
+			-l        list files whose formatting differs from gofumpt's
+			-w        write result to (source) file instead of stdout
+			-extra    enable extra rules which should be vetted by a human
+
+			-lang       str    target Go version in the form \"1.X\" (default from go.mod)
+			-modpath    str    Go module path containing the source file (default from go.mod)
+		with no flags, runs with \"-w .\" by default.
+		$ ./tool fmt          # format current directory and subdirectories.
+		$ ./tool fmt -l ./... # check formatting in current directory and subdirectories."
+
+	'lint [<ARGS>...]'
+	"lint the project using golangci-lint. with no flag, lints the whole project
+		with the config used on the GitHub CI.
+		$ ./tool lint              # lint full project
+		$ ./tool lint cache status # using a golangci-lint subcommand"
+
+	'tidy [<FLAGS>...] [<DIRECTORY>...]'
+	"run \`go mod tidy\` on the given directories.
+		if directories are not given, runs tidy on all directories containing a
+		go.mod file.
+		$ ./tool tidy
+		$ ./tool tidy -v ./contribs/gnodev"
+
+	'tool-anywhere-source'
+	"prints the source of tool-anywhere. tool-anywhere allows you to call tool
+		from any subdirectory.
+		# set up your ~/bin directory if you haven't already
+		$ mkdir -p ~/bin                           # create ~/bin if it doesn't exist
+		$ echo 'export PATH=\"\$PATH:\$HOME/bin\"' >> ~/.profile
+		$ source ~/.profile                        # re-load profile
+		# add tool-anywhere!
+		$ ./tool tool-anywhere-source > ~/bin/tool # save to your ~/bin directory
+		$ chmod +x ~/bin/tool                      # mark as executable
+		$ cd gnovm; tool help                      # success!"
+
+	# TODO: tool run <program> - runs the given program using go run
+	# (ie. with up-to-date codebase), ie. `tool run gno test -v ./...`
+)
+
+# absolute path to script
+tool_root="$(dirname "$(realpath "$0")")"
+
+# list of programs available for `tool install`
+# to add a new program, add it to PROGRAMS, the DOCUMENTATION, and create
+# PROGRAMS_$name.
+PROGRAMS=(
+	'gnokey'
+	'gno'
+	'gnodev'
+	'gnoland'
+	'gnoweb'
+	'gnofaucet'
+	'gnomd'
+	'gnokeykc'
+)
+PROGRAMS_gnokey=(
+	# directory of go.mod
+	'.'
+	# location (relative to go.mod dir)
+	'./gno.land/cmd/gnokey'
+)
+PROGRAMS_gno=(
+	'.'
+	'./gnovm/cmd/gno'
+	# extra go build/install flags
+	"-ldflags -X=github.com/gnolang/gno/gnovm/pkg/gnoenv._GNOROOT=$tool_root"
+)
+PROGRAMS_gnodev=(
+	'./contribs/gnodev'
+	'./cmd/gnodev'
+	"-ldflags -X=github.com/gnolang/gno/gnovm/pkg/gnoenv._GNOROOT=$tool_root"
+)
+PROGRAMS_gnoland=(
+	'.'
+	'./gno.land/cmd/gnoland'
+)
+PROGRAMS_gnoweb=(
+	'.'
+	'./gno.land/cmd/gnoweb'
+)
+PROGRAMS_gnofaucet=(
+	'./contribs/gnofaucet'
+	'.'
+)
+PROGRAMS_gnomd=(
+	'./contribs/gnomd'
+	'.'
+)
+PROGRAMS_gnokeykc=(
```

Comment:
The magic we try to avoid was introduced to avoid having to declare all the binaries, which is a strategy that I find appropriate in the context of contribs/; because contribs/ is designed to be more about exceptions and to avoid having people having to patch .github/workflows and contribs/Makefile each time they want to add a binary.

We can remove Makefile’s magic by only adding inline rules, which is IMO what we currently have in the other Makefiles (contribs/ is an exception if I’m not wrong).